### ISSUE TITLE
loss computation for noisy channel model 

### ISSUE THREAD

**Issue by dorost1234** [shmsw25/Channel-LM-Prompting#7]
Hi
First thanks for being patient with my questions. I have some questions on how the loss for noisy channel model is computed:
1) In `prepro_sentence_pair_single` in data.py you compute token_type_id, do you mind telling me how token_type_id works? I think this is 1 for labels, but I did not fully get it why this is needed and could not follow this function. if we were using direct model, would one still need this function? 
2) In run_model in run.py when you shift the labels and input, is this because you want to use tokens n-1 to predict token n? then you multiply it by the mask in step 1, which I did not get why this is required?

thank you.

**Comment by shmsw25** [comment]
1. We're using it as a label mask when computing the loss [here](https://github.com/shmsw25/Channel-LM-Prompting/blob/master/run.py#L134). Basically, we want to compute the conditional probability of the output only (and not compute the probability of the input) - that's why we have this label mask.
2. Could you tell me with which command do you see 146? It might be when you are using the demonstration method. 128 is the length per example, so when we're using the demonstration method where more than one example is fed into the model, the total length will be larger than 128. (Please see [this line](https://github.com/shmsw25/Channel-LM-Prompting/blob/master/main.py#L137).)
3. Could you point out which line you are referring to? And, regarding BOS/EOS token, I'm not sure which huggingface code you're referring to, but I believe you need to insert BOS/EOS token. It's possible that the code you've been looking at uses high-level API like `batch_encode_plus` which automatically insert BOS/EOS.
4. It has seen set in an arbitrary way to make sure the total input length is smaller than 128 (or whatever the max_input is) after BOS/EOS and verbalizers are being added. There should be a better way to do this, but it anyways prevents the whole sequence to be too long.

**Comment by dorost1234** [comment]
Hi @shmsw25 
Thank you for these great explanations, I understood majority of my questions, I am still trying to understand the loss computation, when you are shifting the labels in [this line](https://github.com/shmsw25/Channel-LM-Prompting/blob/fb9beeff48d580a31d19c642490e701f79d8c3c7/run.py#L127) and logits in [this line](https://github.com/shmsw25/Channel-LM-Prompting/blob/fb9beeff48d580a31d19c642490e701f79d8c3c7/run.py#L119) I am not sure how this works? 

I was wondering how I can get the embeddings for the masked tokens from your formulation? If I use outputs.hidden_states*label_mask, this perhaps shows the label embeddings, but since in the codes, the labels+inputs are fed to the model and there is no [MASK] token, and the model has seen the labels, so this is cheating to obtain the masked tokens embeddings when one multiplies hidden_states with the masks. I am not sure how I can get masked token embeddings,  I appreciate a lot your input on this, thank you.
